src/ansi-c/ansi_c_internal_additions.cpp

```diff
@@ -120,6 +122,25 @@ static std::string architecture_string(T value, const char *s)
          std::string(s) + "=" + std::to_string(value) + ";\n";
 }
 
+using max_alloc_sizet = uint64_t;
+/// The maximum allocation size is determined by the number of bits that
+/// are left in the pointer of width \p pointer_width after allowing for the
+/// signed bit, and the bits used for the objects ID (determined by
+/// \p object_bits).
+/// \param pointer_width: The width of the pointer
+/// \param object_bits : The number of bits used to represent the ID
+/// \return The size in bytes of the maximum allocation supported.
+static max_alloc_sizet
+max_malloc_size(std::size_t pointer_width, std::size_t object_bits)
+{
+  PRECONDITION(pointer_width >= 1);
+  PRECONDITION(object_bits < pointer_width - 1);
+  PRECONDITION(object_bits >= 1);
+  const auto bits_for_offset = pointer_width - object_bits - 1;
+  PRECONDITION(bits_for_offset < std::numeric_limits<max_alloc_sizet>::digits);
+  return ((max_alloc_sizet)1) << (max_alloc_sizet)bits_for_offset;
```

Comment:
I think casting `bits_for_offset` would be unnecessary.

Reply:
I'd prefer to be explicit here, since the specific rules of what type gets promoted are fiddly.